### PR TITLE
feat: version upgrade solution

### LINKED ISSUE #165: version upgrade solution: separate dynamodb(and other datas management) from the cloudformation stack
### Describe the feature

Noticed that new version v.1.1.0 is released, since there is no document to tell how to upgrade, and the dynamodb names keep the same in the cloudformation. Deleting the previous cloudformation stack and creating a new one might be the only way to upgrade.

But if we go this way, all datas in the dynamodb will be lost, including the models we've uploade, inferences we've done etc.

Here's a easy workaround for now (I haven't tested yet):
1. backup all ddb tables
2. delete the stack and create a new onw
3. restore the ddb

For further solution management, it's recommended to separate all datas management from the cloudformation stack, although dynamodb is serverless, we still need to treat it as a database.

### Use Case

version upgrade

### Proposed Solution

Here's my design for version upgrade:
1. Have a script like preflight to check the status of all datas, to see if there's a new ddb table or s3 bucket or something else need to be created.
2. Create a new cloudformation stack and read/write the same data(s3, ddb, etc.), there are some parameters to tell which ddb tables to read/write, also we can have the default table name.
3. After creating the new cloudformation stack, we can test the new version with a new url and token, while the previous version is still online. Just like blue-green deployment.
4. Find a sunny day, change the online url and token to the new version.

### Other Information

Here are some principles(best to follow) for your reference:
1. version upgrade should not destroy current online business
2. version upgrade should not lose any data

### Solution version used

v.1.1.0

### PR DESCRIPTION
**Pull request by Shellmode**
## Description

Refer [README.md](https://github.com/Shellmode/stable-diffusion-aws-extension/blob/main/version_upgrade/README.md)

Fixes #165 (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: version_upgrade/after_run_cf.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    print(f"Restoring S3 bucket: from {src_bucket_name} to {dst_bucket_name}" )
+    subprocess.run(["aws", "s3", "sync", "s3://" + src_bucket_name, "s3://" + dst_bucket_name])
+
+def main():
+    delete_all_tables()
+    restore_all_tables()
+    restore_s3_bucket(src_bucket, dst_bucket)
+
+    all_done = False
+    while all_done == False:
```
Comment (NingLu) [inline version_upgrade/after_run_cf.py:74]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The `==` and `!=` operators use the compared objects&#39; ` __eq__` method to test if they are equal. To check if an object is a singleton, such as `None`, we recommend that you use the `is` identity comparison operator.

[Learn more](https://docs.python.org/3/reference/expressions.html#is)

File: version_upgrade/after_run_cf.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        backups = ddb.list_backups(TableName=ddb_table)
+        for backup in backups['BackupSummaries']:
+            backup_arn = backup['BackupArn']
+            ddb.delete_backup(
+                BackupArn=backup_arn
+            )
+
+def restore_s3_bucket(src_bucket_name, dst_bucket_name):
+    print(f"Restoring S3 bucket: from {src_bucket_name} to {dst_bucket_name}" )
+    subprocess.run(["aws", "s3", "sync", "s3://" + src_bucket_name, "s3://" + dst_bucket_name])
```
Comment (NingLu) [inline version_upgrade/after_run_cf.py:66]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Starting a process with a partial executable path https://bandit.readthedocs.io/en/latest/plugins/b607_start_process_with_partial_path.html

Comment (NingLu) [inline version_upgrade/after_run_cf.py:66]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

subprocess call - check for execution of untrusted input. https://bandit.readthedocs.io/en/latest/plugins/b603_subprocess_without_shell_equals_true.html

**Comment by yike5460** [comment]
Thanks for the feedback, the stack will only need to update instead of deletion if the DynamoDB schema kept unchanged.

**Comment by alvindaiyan** [comment]
it's a good start to me, but we will do this through cfn instead

**Review by alvindaiyan — CHANGES_REQUESTED** [review]
This is a good script for temporary upgrade solution. We will update the cfn in the long run to accept the ddb table and s3 from previous deployment?

**Comment by yike5460** [comment]
Closed for now due to new alternative raised.

**Pull request closed without merging**